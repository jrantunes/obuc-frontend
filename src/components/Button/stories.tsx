import { ComponentMeta, ComponentStory } from "@storybook/react";
import { AiOutlineSave } from "react-icons/ai";
import Button from ".";

export default {
  title: "Button",
  component: Button,
  args: {
    children: "Button",
    size: "medium",
    color: "primary",
    borderRadius: "medium",
    iconPosition: "left",
    hasBoxShadow: false,
    fullWidth: false,
    disabled: false,
    outlined: false,
    minimal: false
  },
  argTypes: {
    icon: {
      control: false
    },
    iconPosition: {
      control: "select"
    },
    as: {
      control: false
    },
    size: {
      control: "select"
    },
    color: {
      control: "select"
    },
    borderRadius: {
      control: "select"
    }
  }
} as ComponentMeta<typeof Button>;

export const Default: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
);

export const withIcon: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
);

withIcon.args = {
  icon: <AiOutlineSave />
};
