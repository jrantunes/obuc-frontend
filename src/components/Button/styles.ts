import styled, { css, DefaultTheme } from "styled-components";
import { darken } from "polished";
import { ButtonProps, ButtonSizes, IconPositions } from ".";

type WrapperProps = {
  hasIcon: boolean;
  hasOnlyIcon: boolean;
} & Pick<
  ButtonProps,
  "size" | "fullWidth" | "minimal" | "color" | "iconPosition"
>;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,

  medium: (theme: DefaultTheme) => css`
    height: 3.6rem;
    font-size: ${theme.font.sizes.large};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};

    svg {
      width: 2.4rem;
      height: 2.4rem;
    }
  `,

  large: (theme: DefaultTheme) => css`
    height: 6rem;
    font-size: ${theme.font.sizes.xlarge};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};

    svg {
      width: 3rem;
      height: 3rem;
    }
  `,

  fullWidth: () => css`
    width: 100%;
  `,

  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};
    transition: color ${theme.transitions.fast};
    padding: 0;
    height: auto;

    &:hover {
      color: ${darken(0.08, theme.colors.primary)};
    }

    &:disabled {
      &:hover {
        background: none;
        color: ${theme.colors.primary};
      }
    }
  `,

  withIcon: (theme: DefaultTheme, iconPosition: IconPositions) => css`
    svg {
      width: 2rem;
      height: 2rem;
      order: ${iconPosition === "right" ? "1" : "0"};

      & + span {
        margin-${iconPosition}: ${theme.spacings.xxsmall};
      }
    }
  `,

  onlyIcon: (theme: DefaultTheme, size: ButtonSizes) => css`
    padding: ${size !== "large"
      ? theme.spacings.xxsmall
      : theme.spacings.small};
    height: auto;
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({
    theme,
    color,
    size,
    fullWidth,
    minimal,
    iconPosition,
    hasIcon,
    hasOnlyIcon
  }) =>
    css`
      display: inline-flex;
      align-items: center;
      justify-content: center;
      background: ${theme.colors[color!]};
      color: ${theme.colors.white};
      font-family: ${theme.font.family};
      font-weight: ${theme.font.regular};
      border: 0;
      border-radius: ${theme.border.radius};
      cursor: pointer;
      padding: ${theme.spacings.xxsmall};
      transition: background-color ${theme.transitions.fast};

      &:hover {
        background: ${minimal ? "none" : darken(0.05, theme.colors[color!])};
      }

      &:disabled {
        opacity: 0.6;
        cursor: not-allowed;

        &:hover {
          background: ${theme.colors[color!]};
        }
      }

      ${hasIcon && wrapperModifiers.withIcon(theme, iconPosition!)}
      ${fullWidth && wrapperModifiers.fullWidth()}
      ${!!size && wrapperModifiers[size!](theme)}
      ${hasOnlyIcon && wrapperModifiers.onlyIcon(theme, size!)}
      ${minimal && wrapperModifiers.minimal(theme)}
    `}
`;
