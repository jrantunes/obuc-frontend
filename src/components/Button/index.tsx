import { ButtonHTMLAttributes, ReactNode } from "react";

import * as S from "./styles";

type ButtonTypes = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonSizes = "small" | "medium" | "large";
export type IconPositions = "left" | "right";

export type ButtonProps = {
  size?: ButtonSizes;
  color?: "primary" | "secondary";
  iconPosition?: IconPositions;
  fullWidth?: boolean;
  minimal?: boolean;
  disabled?: boolean;
  icon?: ReactNode;
} & ButtonTypes;

const Button = ({
  icon,
  children,
  size = "medium",
  color = "primary",
  iconPosition = "left",
  fullWidth = false,
  disabled = false,
  minimal = false,
  ...props
}: ButtonProps) => {
  return (
    <S.Wrapper
      size={size}
      fullWidth={fullWidth}
      minimal={minimal}
      disabled={disabled}
      hasIcon={!!icon}
      color={color}
      iconPosition={iconPosition}
      hasOnlyIcon={!!icon && !children}
      {...props}
    >
      {icon}
      {!!children && <span>{children}</span>}
    </S.Wrapper>
  );
};

export default Button;
