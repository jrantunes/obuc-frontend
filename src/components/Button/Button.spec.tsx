import { screen } from "@testing-library/react";
import { renderWithTheme } from "utils/tests/helpers";

import { AiOutlineSave } from "react-icons/ai";
import Button from ".";

describe("<Button />", () => {
  it("should render with the medium size by default", () => {
    const { container } = renderWithTheme(<Button>Button</Button>);

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyle({
      height: "3.6rem",
      padding: "0.8rem 4.8rem",
      "font-size": "1.8rem"
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with small size", () => {
    renderWithTheme(<Button size="small">Button</Button>);

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyle({
      height: "3rem",
      padding: "0.8rem 3.2rem",
      "font-size": "1.4rem"
    });
  });

  it("should render with large size", () => {
    renderWithTheme(<Button size="large">Button</Button>);

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyle({
      height: "6rem",
      padding: "0.8rem 4.8rem",
      "font-size": "2.0rem"
    });
  });

  it("should render a fullWidth version", () => {
    renderWithTheme(<Button fullWidth>Button</Button>);

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyle({
      width: "100%"
    });
  });

  it("should render a icon version", () => {
    renderWithTheme(
      <Button icon={<AiOutlineSave data-testid="icon" />}>Button</Button>
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render with icon on the right side", () => {
    renderWithTheme(
      <Button icon={<AiOutlineSave data-testid="icon" />} iconPosition="right">
        Button
      </Button>
    );

    expect(screen.getByTestId("icon")).toHaveStyle({
      order: 1
    });
  });

  it("should render a minimal version", () => {
    renderWithTheme(<Button minimal>Button</Button>);

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyle({
      color: "#5ce1e6",
      background: "none"
    });

    expect(screen.getByRole("button", { name: /button/i })).toHaveStyleRule(
      "background",
      "none",
      {
        modifier: ":hover"
      }
    );
  });

  it("should render with a onlyIcon version", () => {
    renderWithTheme(<Button icon={<AiOutlineSave />} />);

    expect(screen.getByRole("button")).toHaveStyle({
      height: "auto"
    });
  });

  it("should render onlyIcon version with custom padding based on large size", () => {
    renderWithTheme(<Button icon={<AiOutlineSave />} size="large" />);

    expect(screen.getByRole("button")).toHaveStyle({
      padding: "2.4rem"
    });
  });

  it("should render onlyIcon version with custom padding based on small/medium size", () => {
    renderWithTheme(<Button icon={<AiOutlineSave />} size="medium" />);

    expect(screen.getByRole("button")).toHaveStyle({
      padding: "0.8rem"
    });
  });
});
