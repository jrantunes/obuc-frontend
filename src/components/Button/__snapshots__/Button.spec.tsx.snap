// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render with the medium size by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #5ce1e6;
  color: #FFFFFF;
  font-family: Roboto,--apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 400;
  border: 0;
  border-radius: 0.4rem;
  cursor: pointer;
  padding: 0.8rem;
  -webkit-transition: background-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out;
  height: 3.6rem;
  font-size: 1.8rem;
  padding: 0.8rem 4.8rem;
}

.c0:hover {
  background: #46dde3;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:disabled:hover {
  background: #5ce1e6;
}

.c0 svg {
  width: 2.4rem;
  height: 2.4rem;
}

<button
  class="c0"
  color="primary"
>
  <span>
    Button
  </span>
</button>
`;
